core/src/main/java/com/sequenceiq/distrox/v1/distrox/service/upgrade/DistroxUpgradeImageSelector.java

```diff
@@ -0,0 +1,88 @@
+                imageInfoV4Response -> runtime.equals(imageInfoV4Response.getComponentVersions().getCdp())).collect(Collectors.toList());
+        validateThereIsMatchingRuntime(upgradeCandidates, imagesWithMatchingRuntime, runtime);
+        ImageInfoV4Response imageInfoV4Response = imagesWithMatchingRuntime.stream().max(ImageInfoV4Response.creationBasedComparator()).orElseThrow();
+        LOGGER.debug("Chosen image {} for {} runtime specified in the request", imageInfoV4Response, runtime);
+        return Optional.of(imageInfoV4Response);
+    }
+
+    private void validateThereIsMatchingRuntime(List<ImageInfoV4Response> upgradeCandidates, List<ImageInfoV4Response> imagesWithMatchingRuntime,
+            String runtime) {
+        boolean hasNoCompatbileImageWithRuntime = imagesWithMatchingRuntime.isEmpty();
+        if (hasNoCompatbileImageWithRuntime) {
+            String availableRuntimes = collectAvailableRuntimes(upgradeCandidates);
+            if (StringUtils.isEmpty(availableRuntimes)) {
+                throw new BadRequestException(String.format("There is no image eligible for the cluster upgrade with runtime: %s.", runtime));
+            } else {
+                throw new BadRequestException(String.format("There is no image eligible for the cluster upgrade with runtime: %s. "
+                        + "Please choose a runtime from the following: %s", runtime, availableRuntimes));
+            }
+        }
+    }
+
+    private String collectAvailableRuntimes(List<ImageInfoV4Response> upgradeCandidates) {
+        return upgradeCandidates.stream()
+                .map(ImageInfoV4Response::getComponentVersions)
+                .map(ImageComponentVersions::getCdp)
+                .distinct()
+                .collect(Collectors.joining(","));
+    }
+
+    private Optional<ImageInfoV4Response> selectRequestedImage(UpgradeV4Request request, List<ImageInfoV4Response> upgradeCandidates) {
+        Optional<ImageInfoV4Response> requestedImage = upgradeCandidates.stream()
+                .filter(candidate -> request.getImageId().equals(candidate.getImageId()))
+                .findFirst();
+        validateImageIsPresent(upgradeCandidates, requestedImage);
+        LOGGER.debug("Chosen image {} as it was specified in the request", requestedImage);
+        return requestedImage;
+    }
+
+    private Optional<ImageInfoV4Response> selectLatestImageFromCandidates(List<ImageInfoV4Response> upgradeCandidates) {
+        ImageInfoV4Response imageInfoV4Response = upgradeCandidates.stream().max(ImageInfoV4Response.creationBasedComparator()).orElseThrow();
+        LOGGER.debug("Choosing latest image {} as either upgrade request is empty or lockComponents is true", imageInfoV4Response);
+        return Optional.of(imageInfoV4Response);
+    }
+
+    private void validateImageIsPresent(List<ImageInfoV4Response> upgradeCandidates, Optional<ImageInfoV4Response> requestedImage) {
```

Comment:
requestedImage.isEmpty() should be the 2nd parameter

Reply:
why? the logic for validation is present right now in `validateImageIsPresent`. If I pull that part out of this method it would bring up part of this logic into the caller.